### DISCUSSION TITLE
Mantine and shadow dom

### DISCUSSION THREAD

**Discussion by dexter4life** [mantinedev/mantine#5224] [Q&A]
I am trying to setup a shadow dom and render a mantine component within this shadow dom. After loading core css into the shadow dom, using <style>...</style> it appears all the variables except those in the :root isn't working.

![image](https://github.com/mantinedev/mantine/assets/10168414/71b71c71-98d5-4163-80e4-eaa127b976d2)

**Comment by rtivital** [comment]
You can change css variables selector with https://mantine.dev/theming/mantine-provider/#cssvariablesselector

**Reply by dexter4life** [reply]
I was able to have those css variables but mostly those in class variables still didn't take effect
![image](https://github.com/mantinedev/mantine/assets/10168414/7d9dd860-a47d-43ea-b63f-1e982d0eb953)

**Reply by dexter4life** [reply]
what can i do to make those accessible too?. Please help

**Reply by dexter4life** [reply]
![image](https://github.com/mantinedev/mantine/assets/10168414/b59d3015-60ab-4215-b00b-378228ad81a0)
Is currently displaying like this. which is different from what i expected

**Reply by rtivital** [reply]
I do not know, I do not have any experience with shadow DOM. You are welcome to research the issue and share your findings with the community.

**Comment by dave-dashbio** [comment]
I think I got this working. In addition to setting `cssVariableSelector` you also need to tell Mantine where the root element is with `getRootElement`. You also need to make sure it injects the portal components inside if the root as well if you want styling for things like tooltips to work.

```typescript
class MyWebComponent extends HTMLElement {
  private readonly root: HTMLElement;

  constructor() {
    super();

    const shadowRoot = this.attachShadow({ mode: "open" });

    const cssUrl = new URL('./styles.css', import.meta.url).href;
    const cssLink = document.createElement('link');
    cssLink.setAttribute('rel', 'stylesheet');
    cssLink.setAttribute('href', cssUrl);
    shadowRoot.appendChild(cssLink);

    this.root = document.createElement('div');
    this.root.setAttribute('id', 'root');
    shadowRoot.appendChild(this.root);
  }

  connectedCallback() {
    const theme = createTheme({
      components: {
        Portal: {
          defaultProps: {
            target: this.root
          },
        }
      }
    });

    ReactDOM.createRoot(this.root).render(
      <MantineProvider
        cssVariablesSelector="#root"
        getRootElement={() => this.root}
        theme={theme}
      >
        <ModalsProvider>
          <App />
        </ModalsProvider>
      </MantineProvider>,
    );
  }
}

customElements.define("my-component", MyWebComponent);
```

**Reply by khash** [reply]
Is this not really inefficient as we are wrapping each instance of this component in a `MantineProvider`? Is there a way to find the provider at the root of the app and create portal instead?
